server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -125,79 +124,152 @@ public SearchHit(int docId, String id) {
+        this.version = version;
+        this.seqNo = seqNo;
+        this.primaryTerm = primaryTerm;
+        this.source = source;
+        this.highlightFields = highlightFields;
+        this.sortValues = sortValues;
+        this.matchedQueries = matchedQueries;
+        this.explanation = explanation;
+        this.shard = shard;
+        this.index = index;
+        this.clusterAlias = clusterAlias;
+        this.innerHits = innerHits;
+        this.documentFields = documentFields;
+        this.metaFields = metaFields;
     }
 
-    public SearchHit(StreamInput in) throws IOException {
-        docId = -1;
-        score = in.readFloat();
+    public static SearchHit readFrom(StreamInput in) throws IOException {
+        final float score = in.readFloat();
+        final int rank;
         if (in.getTransportVersion().onOrAfter(TransportVersions.V_8_8_0)) {
             rank = in.readVInt();
+        } else {
+            rank = NO_RANK;
         }
-        id = in.readOptionalText();
+        final Text id = in.readOptionalText();
         if (in.getTransportVersion().before(TransportVersions.V_8_0_0)) {
             in.readOptionalText();
         }
-        nestedIdentity = in.readOptionalWriteable(NestedIdentity::new);
-        version = in.readLong();
-        seqNo = in.readZLong();
-        primaryTerm = in.readVLong();
-        source = in.readBytesReference();
+        final NestedIdentity nestedIdentity = in.readOptionalWriteable(NestedIdentity::new);
+        final long version = in.readLong();
+        final long seqNo = in.readZLong();
+        final long primaryTerm = in.readVLong();
+        BytesReference source = in.readBytesReference();
         if (source.length() == 0) {
             source = null;
         }
+        Explanation explanation = null;
         if (in.readBoolean()) {
             explanation = readExplanation(in);
         }
-        documentFields.putAll(in.readMap(DocumentField::new));
-        metaFields.putAll(in.readMap(DocumentField::new));
-
-        int size = in.readVInt();
-        if (size == 0) {
-            highlightFields = emptyMap();
-        } else if (size == 1) {
-            HighlightField field = new HighlightField(in);
-            highlightFields = singletonMap(field.name(), field);
-        } else {
-            Map<String, HighlightField> highlightFields = new HashMap<>();
-            for (int i = 0; i < size; i++) {
-                HighlightField field = new HighlightField(in);
-                highlightFields.put(field.name(), field);
-            }
-            this.highlightFields = unmodifiableMap(highlightFields);
-        }
-
-        sortValues = new SearchSortValues(in);
+        final Map<String, DocumentField> documentFields = in.readMap(DocumentField::new);
+        final Map<String, DocumentField> metaFields = in.readMap(DocumentField::new);
+        final Map<String, HighlightField> highlightFields = in.readMapValues(HighlightField::new, HighlightField::name);
```

Comment:
Could you make this an unmodifiable map like it was before?

Reply:
I did that at the bottom of the method :)